cpp/src/io/json/nested_json_gpu.cu

```diff
@@ -0,0 +1,640 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "nested_json.hpp"
+
+#include <io/fst/logical_stack.cuh>
+#include <io/fst/lookup_tables.cuh>
+#include <io/utilities/hostdevice_vector.hpp>
+
+#include <cudf/types.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/exec_policy.hpp>
+
+namespace cudf::io::json {
+
+//------------------------------------------------------------------------------
+// JSON-TO-STACK-OP DFA
+//------------------------------------------------------------------------------
+namespace to_stack_op {
+
+// Type used to represent the target state in the transition table
+using StateT = char;
+
+/**
+ * @brief Definition of the DFA's states
+ */
+enum DFA_STATES : StateT {
+  // The state being active while being outside of a string. When encountering an opening bracket
+  // or curly brace, we push it onto the stack. When encountering a closing bracket or brace, we
+  // pop from the stack.
+  TT_OOS = 0U,
+
+  // The state being active while being within a string (e.g., field name or a string value). We do
+  // not push or pop from the stack while being in this state.
+  TT_STR,
+
+  // The state being active after encountering an escape symbol (e.g., '\'), while being in the
+  // TT_STR state.
+  TT_ESC,
+
+  // Total number of states
+  TT_NUM_STATES
+};
+
+/**
+ * @brief Definition of the symbol groups
+ */
+enum class DFASymbolGroupID : uint32_t {
+  OpenBrace,         ///< Opening brace SG: {
+  OpenBracket,       ///< Opening bracket SG: [
+  CloseBrace,        ///< Closing brace SG: }
+  CloseBracket,      ///< Closing bracket SG: ]
+  Quote,             ///< Quote character SG: "
+  Escape,            ///< Escape character SG: '\'
+  Other,             ///< SG implicitly matching all other characters
+  NUM_SYMBOL_GROUPS  ///< Total number of symbol groups
+};
+
+// The i-th string representing all the characters of a symbol group
+const std::vector<std::string> symbol_groups = {"{", "[", "}", "]", "\"", "\\"};
+
+// Transition table
+const std::vector<std::vector<StateT>> transition_table = {
+  /* IN_STATE         {       [       }       ]       "       \    OTHER */
+  /* TT_OOS    */ {TT_OOS, TT_OOS, TT_OOS, TT_OOS, TT_STR, TT_OOS, TT_OOS},
+  /* TT_STR    */ {TT_STR, TT_STR, TT_STR, TT_STR, TT_OOS, TT_ESC, TT_STR},
+  /* TT_ESC    */ {TT_STR, TT_STR, TT_STR, TT_STR, TT_STR, TT_STR, TT_STR}};
+
+// Translation table (i.e., for each transition, what are the symbols that we output)
+const std::vector<std::vector<std::vector<char>>> translation_table = {
```

Comment:
Why is this a triply nested vector? It seems like it only needs two levels to act as a lookup table.

Reply:
That is because the "translated" output for a a transition can be more than a single symbol. 
`translation_table[STATE][SYMBOL_GROUP_ID] => output_string`
E.g., we have two cases where we output 2 tokens on one transition:
[Translation Table (see sheet "translation table" state:LON, sgid:'\]'](https://docs.google.com/spreadsheets/d/1G0Nc0Fj9c000vtGs0c0SOTmfyMiMxnVFQB_TDFBO3AU/edit?usp=sharing)